[SR-15691] Invalid warning: 'self' refers to the method 'Foo.self'

|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-15691      |
   |Radar             | None         |
   |Original Reporter | @keith      |
   |Type              | Bug    |

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       |     |
   |Labels            | Bug        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: 92266992a8216159c3017c1eaba44660

  </details>





**Issue Description:**


With this code and Swift 5.5, it compiled fine:

``` none
enum Foo {
    case `self`

    init() {
        self = .self
    }
}
```

With the 2021-12-23 main snapshot (org.swift.50202112231a) compiling gives an seemingly impossible to silence warning, here are the variations I tried:

``` none
enum Foo {
    case `self`

    init() {
        // This option produces:
        // /tmp/qux.swift:5:17: warning: 'self' refers to the method 'Foo.self', which may be unexpected
        //         self = .self
        //                 ^
        // /tmp/qux.swift:5:17: note: use 'Foo.self' to silence this warning
        //         self = .self
        //                 ^
        //                 Foo.
        self = .self

        // This option fails to compile
        // /tmp/qux.swift:15:16: error: cannot assign value of type 'Foo.Type' to type 'Foo'
        //         self = Foo.self
        //                ^~~~~~~~
        // self = Foo.self

        // This option produces:
        // /tmp/qux.swift:22:20: warning: 'self' refers to the method 'Foo.self', which may be unexpected
        //         self = Foo.`self`
        //                    ^
        // /tmp/qux.swift:22:20: note: use 'Foo.self' to silence this warning
        //         self = Foo.`self`
        //                    ^
        //                    Foo.
        self = Foo.`self`
    }
}
```

I believe this might have been introduced as the fix for <https://bugs.swift.org/browse/SR-4559> here <https://github.com/apple/swift/commit/7e0a7f322aedd93ae7359884f7c8b4e6884f63d9>

cc jackmarch (JIRA User)

--- a/lib/Sema/MiscDiagnostics.cpp
+++ b/lib/Sema/MiscDiagnostics.cpp
@@ -4748,19 +4748,20 @@ static void diagUnqualifiedAccessToMethodNamedSelf(const Expr *E,
           if (auto typeContext = DC->getInnermostTypeContext()) {
             // self() is not easily confusable
             if (!isa<CallExpr>(Parent.getAsExpr())) {
-
               auto baseType = typeContext->getDeclaredInterfaceType();
-              auto baseTypeString = baseType.getString();
+              if (!baseType->getEnumOrBoundGenericEnum()) {
+                auto baseTypeString = baseType.getString();
 
-              Ctx.Diags.diagnose(E->getLoc(), diag::self_refers_to_method,
-                                 baseTypeString);
+                Ctx.Diags.diagnose(E->getLoc(), diag::self_refers_to_method,
+                    baseTypeString);
 
-              Ctx.Diags
+                Ctx.Diags
                   .diagnose(E->getLoc(),
-                            diag::fix_unqualified_access_member_named_self,
-                            baseTypeString)
+                      diag::fix_unqualified_access_member_named_self,
+                      baseTypeString)
                   .fixItInsert(E->getLoc(), diag::insert_type_qualification,
-                               baseType);
+                      baseType);
+              }
             }
           }
         }

--- a/test/Parse/self_rebinding.swift
+++ b/test/Parse/self_rebinding.swift
@@ -116,3 +116,13 @@ struct TypeWithSelfProperty {
     
     let `self`: () = ()
 }
+
+enum EnumCaseNamedSelf {
+    case `self`
+
+    init() {
+        self = .self // OK
+        self = .`self` // OK
+        self = EnumCaseNamedSelf.`self` // OK
+    }
+}

